File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(str string) int {
    start, end, mapp, maxLen := 0, 0, make(map[byte]int), 0
    for end < len(str) {
        if val, ok := mapp[str[end]]; !ok || val < start {
            mapp[str[end]] = end
            maxLen = max(maxLen, end - start + 1)
            end++
        } else {
            start = val + 1
        }
    }
    return maxLen
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(str) {
    let start = 0, end = 0, maxLen = 0, mapp = {};
    while (end < str.length) {
        if (!mapp.hasOwnProperty(str[end]) || mapp[str[end]] < start) {
            mapp[str[end]] = end;
            maxLen = Math.max(maxLen, end - start + 1);
            end++;
        } else {
            start = mapp[str[end]] + 1;
        }
    }
    return maxLen;
};